### ISSUE TITLE
Router.getUrl() returns inconsistent values

### ISSUE THREAD

**Issue by Wnt** [vaadin/flow#3519]
When route is registered with
```java
@Route(value = "customer", layout = MainView.class)
public class CustomerDetailsRoute extends Composite<CustomerDetailsView>
    implements HasUrlParameter<String> {

    @Override
    public void setParameter(BeforeEvent event, @WildcardParameter String parameter) {
```
`Router.getUrl(CustomerDetailsRoute.class)` returns
`customer`

However when route is registered with
```java
@Route(value = "customer", layout = MainView.class)
public class CustomerDetailsRoute extends Composite<CustomerDetailsView>
    implements HasUrlParameter<Long>
```
Router.getUrl() returns
`customer/{Long}`

Make return values more consistent.

@Legioth suggested:

> It might actually be good to have even three different methods:
> * `getUrl` that throws if a parameter is required
> * `getUrlBase` that returns the static part of the URL without any parameter prefixes
> * `getUrlPattern` that returns something similar to the current `customers/{Long}` syntax

**Comment by Wnt** [comment]
The usecase where I was trying to use the `customer`part :

```java
grid.addColumn(TemplateRenderer.<Customer>of
		("<a router-link href=\"" + url + "/[[item.id]]\">Inspect [[item.name]]</a>")
		.withProperty("name", Customer::getFirstName)
		.withProperty("id", Customer::getId));

```

**Comment by Wnt** [comment]
Looks like the Router.getUrl() returns
just `customer` only when there is a `@WildcardParameter` annotation on the setParameter method's String parameter

**Comment by caalador** [comment]
Yes the type should not matter for the return value, but if you have the url parameter either as `@OptionalParameter` or `@WildcardParameter` then you will get with `getUrl(Class)` the route url without the parameter part as it's not needed.

For a required parameter we can't know what to do with the parameter part so we just give what we use when handling a with parameter request.

getUrl and getUrlBase are not problematic, but the case of getUrlPattern is a bit ambiguous, should it return the value type for optional and wildcard?
Should optional/wildcard be specified somehow?

**Comment by Wnt** [comment]
From my point of view, making getUrl throw on non optional parameters and adding getUrlBase would already make the API much better.

**Comment by Legioth** [comment]
I agree. `getUrlPattern` can be considered separately.

**Comment by Wnt** [comment]
Looks like this was fixed via #3523
